### DISCUSSION TITLE
Create React App with FastAPI backend (running on Railway.app) works locally but not when deployed to Vercel (304 & 405 responses)

### DISCUSSION THREAD

**Discussion by chrisLoweDev** [vercel/community#2234] [Help] [answered]
### Summary

Hi, hopefully this is going to be straight forward for someone, but I'm struggling with it...

My Create React App works fine locally, communicating with a FastAPI backend and DB on railway.app without issue. I have set the "proxy": "https://<my-railway-domain>" in package.json as the root to which all my http requests go within my app.

However, when I upload to Vercel, I just get 304 responses for most of the api calls and then, e.g. when I try to login on my homepage I get a 405. I can see from the logs in Railway that there are no interactions happening, so the requests are not making it that far. I'm assuming there's another setting somewhere that I need to change that directs the requests to my backend, but I can't find what that is.

Any advice would be appreciated and sorry if I've missed something obvious from the docs here.
Thanks,
Chris

**Comment by maddsua** [comment]
Hi @chrisLoweDev ! Could you try to reproduce the issue on a dummy project that would just request the Railway's API without any additional logic? Or maybe try "pulling" some other public API to see on which side the problem is. To me, it seems like a strange behavior of CORS, although it should not interfere in your case at all.

**Reply by chrisLoweDev** [reply]
Ok thanks @maddsua.
Will try and have a look at this later on today and report back.

**Reply by chrisLoweDev** [reply]
Hey @maddsua, sorry for the delay (busy week)! I've added the proxy url for my railway server into the package.json in just a basic CRA template and deployed that within vercel [here](https://create-react-app-nu-flax-65.vercel.app/). That seems to have no issue getting a response, so I'm struggling to recreate the issue. It must be something I've done within my app though, some setting somewhere 😖

It does seem like a CORS thing though, as you say, so I'll keep searching for the issue. If anything else springs to mind I'd be happy to hear it.

Thanks again

**Reply by chrisLoweDev** [reply]
After a bit more playing and searching, I've got the two servers talking, by setting up some CORS Middleware in my FastAPI app ([based on this](https://fastapi.tiangolo.com/tutorial/cors/)), and using the full backend URL in my `fetch` requests in the frontend, but still no joy using the "proxy" setting in package.json. This is really no surprise, however, since I stumbled across [this](https://create-react-app.dev/docs/proxying-api-requests-in-development/#:~:text=Keep%20in%20mind%20that%20proxy%20only%20has%20effect%20in%20development%20(with%20npm%20start)%2C%20and%20it%20is%20up%20to%20you%20to%20ensure%20that%20URLs%20like%20/api/todos%20point%20to%20the%20right%20thing%20in%20production), which states that the "proxy" setting only works when using `npm start` and not, as I need for production, `npm run build`. There seems to be a lot of confusion around whether or not this is the case, but it certainly seems to be for me.

Therefore, to avoid needing the full URL in all of my HTTP requests (not ideal for development), I need another way of connecting up the front and backend using a proxy. Assuming either NGINX or Apache would be good for this, unless anyone knows of a more straight forward way to do it? Perhaps there's something just in the settings of Vercel that would all my requests to get through?

Again, any help is appreciated.

**Reply by chrisLoweDev** [reply]
I feel like this is an issue that _every_ app with a separately deployed frontend and backend will face, no? The fact that I'm struggling to find any clear guidance on how to do it suggests I'm doing something silly/stupid/unnecessary.

Should I not have my project set up in the way I've got it? I.e. Python FastAPI backend logic on Railway and React UI on Vercel? This struck me as a logical architecture...

**Reply by maddsua** [reply]
> I feel like this is an issue that _every_ app with a separately deployed frontend and backend will face, no? The fact that I'm struggling to find any clear guidance on how to do it suggests I'm doing something silly/stupid/unnecessary.
> 
> Should I not have my project set up in the way I've got it? I.e. Python FastAPI backend logic on Railway and React UI on Vercel? This struck me as a logical architecture...

Oh this issue is just spiraling up in complexity. 🤯
As I get it, you just want to hide the DB's endpoint using Vercel's redirects? If so, you might also try creating a subdomain for the DB and assigning it to the Railway backend. ~~Or just move to Netlify, I had virtually no issues doing stuff like that there~~


BTW, where are the Vercel's support engineers when we need them the most? @amyegan 

An afterthought about long URLs. I use Firebase storage in one of my projects, and it has awfully long URLs, not ideal. So me, being a lazyass, I just threw in a piece of code that abstracts it away. Keep the demons in a box or smth i dont read dark fantasy 😂

**Reply by maddsua** [reply]
> I feel like this is an issue that _every_ app with a separately deployed frontend and backend will face, no? The fact that I'm struggling to find any clear guidance on how to do it suggests I'm doing something silly/stupid/unnecessary.
> 
> Should I not have my project set up in the way I've got it? I.e. Python FastAPI backend logic on Railway and React UI on Vercel? This struck me as a logical architecture...

I mean, we clearly have a bit of miscommunication here, but it's okay.
Well yes, if you want to do a request for a server, that is not on the same domain as the frontend - you're gonna face CORS. I assume that you already figured it out, if you got them talking to each other.
I have no experience working with your stack, but it should not be that frustrating to just make that requests work.

Actually you might try to do the redirects manually with Edge Functions kinda like this:

- You have an API route on the frontend: `/api/edge/dblink` (this is what I would name it)
- There is an Edge Function that just relays all the requests to the Railway. It's possible that it won't face CORS at all, bc it's not a browser and it might just not care, but I'm not 100% sure about that
- Should work as a charm

**Reply by chrisLoweDev** [reply]
Ok thanks @maddsua - appreciate your help on this and regarding the confusion around the issue I'm having, that is 100% down to me 🙈  sorry! On some of your points:

> you just want to hide the DB's endpoint using Vercel's redirects? If so, you might also try creating a subdomain for the DB and assigning it to the Railway backend.

Nothing here is related to the DB. My DB is all setup alongside my API on Railway, and those two are getting on just fine 😝 Only my backend API interacts with the DB, everything in react goes via the API in order to get/post/put stuff to the DB. 

> ~~Or just move to Netlify, I had virtually no issues doing stuff like that there~~

Ha, I'm actually one step ahead of you on this and got that set up yesterday, but alas same issue there (tbh that's not a surprise since 
up until recently I've been trying to get the package.json to handle my redirects!). That's not to say the solution is not more straight forward to solve on Netlify, so I'd welcome any thoughts on that too.

> you're gonna face CORS. I assume that you already figured it out, if you got them talking to each other.

Yep, CORS is all configured in my API and all seems fine. When I use the full backend URL in my react http requests, there's no issues and everything works as expected. Again, though, I want to avoid needing to have the full BE URL in all of my React components.

> Actually you might try to do the redirects manually with Edge Functions kinda like this:...

Ok will do, I'll have a look at Edge functions, and will also look at Vercel's redirects and rewrites, as I read somewhere that they might help

Hopefully, once this is sorted, I'll look back on this conversation and laugh at how trivial the solution eventually was 😅 
Thanks

**Reply by chrisLoweDev** [reply, marked as answer]
@maddsua we're up and running 🤗 The missing piece of the puzzle was indeed the rewrites, specified in a vercel.json file. Now, any HTTP request within my react app that begins with "/api" is forwarded my backend server.

```
{
    "rewrites": [
      {
        "source": "/api/:path*",
        "destination": "<insert_my_railway_url>/api/:path*"
      }
    ]
  }
```

For the record, I've removed the CORS middleware that I had in and we're _still_ working, which is a little odd and, arguably, a slight concern... I would have expected my vercel app to be required as a defined origin still. Although, perhaps if the rewrite is actually an edge function under the hood, then it's just being allowed in. I'm sure I can add protection in there, if necessary at some point.

Thanks for the help,
C

**Reply by maddsua** [reply]
> @maddsua we're up and running 🤗 The missing piece of the puzzle was indeed the rewrites, specified in a vercel.json file. Now, any HTTP request within my react app that begins with "/api" is forwarded my backend server.
> 
> ```
> {
>     "rewrites": [
>       {
>         "source": "/api/:path*",
>         "destination": "<insert_my_railway_url>/api/:path*"
>       }
>     ]
>   }
> ```
> 
> For the record, I've removed the CORS middleware that I had in and we're _still_ working, which is a little odd and, arguably, a slight concern... I would have expected my vercel app to be required as a defined origin still. Although, perhaps if the rewrite is actually an edge function under the hood, then it's just being allowed in. I'm sure I can add protection in there, if necessary at some point.
> 
> Thanks for the help, C

Glad to know that you were able to sort that out!
Now, about CORS. You don't really have to deal with it when communicating server-to-server. If you wish to limit the access, you need to check the Origin header and reject the request if it does not match with your frontend domain. Redirects on Vercel are a bit weird =)

**Reply by amyegan** [reply]
I'm glad y'all were able to solve this one together! Last week was very busy with [Vercel Ship](https://vercel.com/ship) feature announcements, so I really appreciate you collaborating to find a solution before I was able to respond 🙇
